### ISSUE TITLE
Exception 

### ISSUE THREAD

**Issue by gpizo** [jcachat/EntityFramework.DynamicFilters#74]
Hi there:

As you requested this is new issue with all the details. Model, query, filter and error. I hope it's clear, if not please let me know. Thanks in advance for your help.

[Table(TableNames.EntityA)]
    public abstract class EntityA : DataEntityBase, IOriginAware 
    {
        protected EntityA()
        {
            OriginId = Enumerations.Origin1;
        }

```
    [StringLength(Constraints.IdentifierLength), Index(IsUnique = true)]
    public string Identifier { get; set; }

    [ForeignKey(ForeignKeyNames.Origin)]
    public Origin OriginId { get; set; }
    public virtual OriginLookup Origin { get; set; }

}

[Table(TableNames.EntityB)]
public class EntityB : EntityA
{
    public string InterstateNumber { get; set; }
    public virtual ICollection<FirearmLicenceFirearm> FirearmLicenceFirearms { get; set; }
}

public class EntityC : DataEntityBase, IOriginAware
{
    [StringLength(Constraints.IdentifierLength), Index(IsUnique = true)]
    public string Identifier { get; set; }

    [ForeignKey(ForeignKeyNames.Origin)]
    public Origin OriginId { get; set; }
    public virtual OriginLookup Origin { get; set; }

}

public class ManyToManyEntity : DataEntityBase
{
    [ForeignKey(ForeignKeyNames.EntityB)]
    public int EntityBId { get; set; }
    public virtual EntityB EntityB { get; set; }

    [ForeignKey(ForeignKeyNames.EntityC)]
    public int EntityCId { get; set; }
    public virtual EntityC EntityC { get; set; }

    public virtual ICollection<Condition> Conditions { get; set; }

    public Period ValidityPeriod { get; set; }
}
```

My Filter:

modelBuilder.Filter(FilterNames.Origing1Only, (IOriginAware e) => e.OriginId == Origin1);

```
Query: 
var result = new SearchResult();
var query =(from f in _manytomany select f.EntityB);
result.TotalCount = query.Count(); -- > Breaks

var result = new SearchResult();
var query =(from f in _manytomany select f.EntityC);
result.TotalCount = query.Count(); -- > Works
```

Parameter name: targetSet
at System.Data.Entity.Utilities.Check.NotNullT
at System.Data.Entity.Core.Common.CommandTrees.ExpressionBuilder.DbExpressionBuilder.Scan(EntitySetBase targetSet)
at EntityFramework.DynamicFilters.DynamicFilterQueryVisitorCSpace.Visit(DbPropertyExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DbPropertyExpression.AcceptTResultType
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbPropertyExpression expression)
at EntityFramework.DynamicFilters.DynamicFilterQueryVisitorCSpace.Visit(DbPropertyExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DbPropertyExpression.AcceptTResultType
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitListTElement
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionList(IList1 list) at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbNewInstanceExpression expression) at System.Data.Entity.Core.Common.CommandTrees.DbNewInstanceExpression.Accept[TResultType](DbExpressionVisitor1 visitor)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitListTElement
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionList(IList1 list) at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbNewInstanceExpression expression) at System.Data.Entity.Core.Common.CommandTrees.DbNewInstanceExpression.Accept[TResultType](DbExpressionVisitor1 visitor)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbProjectExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DbProjectExpression.AcceptTResultType
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionBinding(DbExpressionBinding binding)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionBindingEnterScope(DbExpressionBinding binding)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbSkipExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DbSkipExpression.AcceptTResultType
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbLimitExpression expression)
at System.Data.Entity.Core.Common.CommandTrees.DbLimitExpression.AcceptTResultType
at EntityFramework.DynamicFilters.DynamicFilterInterceptor.TreeCreated(DbCommandTreeInterceptionContext interceptionContext)
at System.Data.Entity.Infrastructure.Interception.DbCommandTreeDispatcher.b__0(IDbCommandTreeInterceptor i, DbCommandTreeInterceptionContext c)
at System.Data.Entity.Infrastructure.Interception.InternalDispatcher1.Dispatch[TInterceptionContext,TResult](TResult result, TInterceptionContext interceptionContext, Action2 intercept)
at System.Data.Entity.Infrastructure.Interception.DbCommandTreeDispatcher.Created(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
at System.Data.Entity.Core.Common.DbProviderServices.CreateCommandDefinition(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.CreateCommandDefinition(ObjectContext context, DbQueryCommandTree tree)
--- End of inner exception stack trace ---
at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.CreateCommandDefinition(ObjectContext context, DbQueryCommandTree tree)
at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.Prepare(ObjectContext context, DbQueryCommandTree tree, Type elementType, MergeOption mergeOption, Boolean streaming, Span span, IEnumerable1 compiledQueryParameters, AliasGenerator aliasGenerator) at System.Data.Entity.Core.Objects.ELinq.ELinqQueryState.GetExecutionPlan(Nullable1 forMergeOption)
at System.Data.Entity.Core.Objects.ObjectQuery1.<>c__DisplayClass7.<GetResults>b__6() at System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
at System.Data.Entity.Core.Objects.ObjectQuery1.<>c__DisplayClass7.<GetResults>b__5() at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute[TResult](Func1 operation)
at System.Data.Entity.Core.Objects.ObjectQuery1.GetResults(Nullable1 forMergeOption)
at System.Data.Entity.Core.Objects.ObjectQuery1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__0() at System.Data.Entity.Internal.LazyEnumerator1.MoveNext()
at System.Collections.Generic.List1..ctor(IEnumerable1 collection)
at System.Linq.Enumerable.ToListTSource

**Comment by jcachat** [comment]
What is the type "Origin"?  And where does the "Origin1" value come from in your filter expression?

**Comment by gpizo** [comment]
Origin is an enumeration:

 public enum Origin
    {
        [Lookup("Origin1")]
        Origin1= 1,
        [Lookup("Origin2")]
        Origin2= 2,
        [Lookup("Origin3")]
        Origin3= 3
    }

**Comment by jcachat** [comment]
I'm not able to reproduce the problem.  Can you try to put this together into a standalone VS project that demonstrates the problem and send it to me?

**Comment by DanielGrandis** [comment]
Hi gpizo, have you found a solution to the problem? I've cloned repository and debugged source code. Exception was thrown at DyanimicFilterQueryVisitorCSpace.cs, row 127 because it can't find a EntitySet with the name of the derived class, infact EF create just an EntitySet for the base class. The row 128 try to call DbExpressionBuider.Scan(entitySet) on a null value, and that causes the Exception.

by now, I've workaround this with this:

`   if (entitySet == null) return baseResult;`
but, obiouvsly, this cause not filter applied at all on the derived type.

I've also tried to switch the EntitySet to the base Type, but this cause another exception: isn't possible to scan expression for a different type (the base type) other than the initial derived type passed in to the expression visitor.

I'm on fire, but I still not finding any solution.

Have you some interesting idea on that?

Thanks in advance!
Daniel
